### PR TITLE
[HttpFoundation] Find the original request protocol version

### PR DESCRIPTION
**Pull request by thewilkybarkid**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21415
| License       | MIT
| Doc PR        | 

Adds a new method to `Request` to find the original protocol version from the `Via` header, if the request is from a trusted proxy.

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 1]
```diff
@@ -1536,6 +1536,24 @@ public function isMethodCacheable()
...
 
     /**
+     * Returns the original protocol version.
+     *
+     * @return string
+     */
+    public function getOriginalProtocolVersion()
+    {
+        if ($this->isFromTrustedProxy()) {
+            preg_match('~^(HTTP/)?([1-9].[0-9]) ~', $this->headers->get('Via'), $matches);
```
Comment (dunglas) [inline src/Symfony/Component/HttpFoundation/Request.php]: The `.` must be escaped (currently it matches `HTTP/1^1 ` for instance).

**Review by dunglas — CHANGES_REQUESTED** [review]

**Comment by nicolas-grekas** [comment]
The other methods that use `isFromTrustedProxy` don't have the "Original" keyword in their name. What about just `getProtocolVersion()`?
Is there anything else that could be relevant in the "Via" header? Do you have any link documenting its value?

**Comment by nicolas-grekas** [comment]
Status: needs work

**Comment by nicolas-grekas** [comment]
Security issue: why should "Via" be trusted?
Any RFC that defines the semantic of the header?

**Comment by dunglas** [comment]
@nicolas-grekas https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.45

**Comment by thewilkybarkid** [comment]
@nicolas-grekas I used the term 'original' to distinguish the request that the user originally made (to some proxy) with the one that PHP has received. Happy to rename if there's a more suitable term.

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 2]
```diff
@@ -1536,6 +1536,24 @@ public function isMethodCacheable()
     }
 
     /**
+     * Returns the original protocol version.
```
Comment (javiereguiluz) [inline src/Symfony/Component/HttpFoundation/Request.php]: This description should be improved to explain very briefly what does "the original protocol" mean. Thanks!

**Comment by nicolas-grekas** [comment]
We don't make any distinction for "original" host/ip/proto/port. I don't see why this is different for the version of the proto.

**Comment by thewilkybarkid** [comment]
The response should be the protocol version used by the proxy, not the original client, so they need to be kept separate.

**Comment by dunglas** [comment]
Maybe can we add a method to parse the `Via` header, with a shortcut to only retrieve the value of the first (original) protocol and version used?

It may be interesting to know the original protocol version (for instance to add different `Link` HTTP headers depending of the client capabilities).

**Comment by nicolas-grekas** [comment]
> The response should be the protocol version used by the proxy, not the original client, so they need to be kept separate.

that's exactly my point yes: getHost, isSecure, getPort all return the info as used by the client - not the proxy. This should be the same here to me.
If one want to get access to the protocol used by the proxy, just call `$request->server->get('SERVER_PROTOCOL')`.

**Review by nicolas-grekas — APPROVED** [review]
:+1:

**Comment by fabpot** [comment]
:-1: Your application is talking to the proxy, so we should return the HTTP protocol of the proxy, not the one used between the client and the proxy. I understand the use case, but that should be done in your application, not in Symfony.

**Comment by thewilkybarkid** [comment]
The protocol version isn't being changed in the response.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @thewilkybarkid.

**Pull request closed without merging**